lua/ttt2/libraries/modelbuilder.lua

```diff
@@ -0,0 +1,316 @@
+-- @param Entity viewModel The view model of the weapon
+-- @realm client
+function modelbuilder.ResetBonePositions(viewModel)
+	local boneCount = viewModel:GetBoneCount()
+
+	if not boneCount then return end
+
+	for i = 0, boneCount do
+		viewModel:ManipulateBoneScale(i, Vector(1, 1, 1))
+		viewModel:ManipulateBoneAngles(i, Angle(0, 0, 0))
+		viewModel:ManipulateBonePosition(i, Vector(0, 0, 0))
+	end
+end
+
+---
+-- Updates the bone positions of the view model of a weapon. Is used to hide a
+-- default view model of a weapon.
+-- @param Entity wep The weapon for which the view model should be updated
+-- @param Entity viewModel The view model of the weapon
+-- @realm client
+function modelbuilder.UpdateBonePositions(wep, viewModel)
+	if not wep.ViewModelBoneMods then
+		modelbuilder.ResetBonePositions(viewModel)
+
+		return
+	end
+
+	local allBones = {}
+
+	for i = 0, viewModel:GetBoneCount() do
+		local bonename = viewModel:GetBoneName(i)
+
+		if wep.ViewModelBoneMods[bonename] then
+			allBones[bonename] = wep.ViewModelBoneMods[bonename]
+		else
+			allBones[bonename] = {
+				scale = Vector(1, 1, 1),
+				pos = Vector(0, 0, 0),
+				angle = Angle(0, 0, 0)
+			}
+		end
+	end
+
+	for identifier, dataTable in pairs(allBones) do
+		local bone = viewModel:LookupBone(identifier)
+
+		if not bone then continue end
+
+		local s = Vector(dataTable.scale.x, dataTable.scale.y, dataTable.scale.z)
+		local p = Vector(dataTable.pos.x, dataTable.pos.y, dataTable.pos.z)
+		local ms = Vector(1, 1, 1)
+
+		local cur = viewModel:GetBoneParent(bone)
+
+		while (cur >= 0) do
+			local pScale = allBones[viewModel:GetBoneName(cur)].scale
+
+			ms = ms * pScale
+			cur = viewModel:GetBoneParent(cur)
+		end
+
+		s = s * ms
+
+		if viewModel:GetManipulateBoneScale(bone) ~= s then
+			viewModel:ManipulateBoneScale(bone, s)
+		end
+
+		if viewModel:GetManipulateBoneAngles(bone) ~= dataTable.angle then
+			viewModel:ManipulateBoneAngles(bone, dataTable.angle)
+		end
+
+		if viewModel:GetManipulateBonePosition(bone) ~= p then
+			viewModel:ManipulateBonePosition(bone, p)
+		end
+	end
+end
+
+---
+-- Builds a render order or returns the cache if already built. The render order is necessary
+-- because models have to be drawn before sprites and quads.
+-- @param table elements the view elements table
+-- @param table cachedRenderOrder The cached render order that is uesed as a fallback
+-- @return table Returns a new or the cached render order
+-- @realm client
+function modelbuilder.BuildRenderOrder(elements, cachedRenderOrder)
+	if cachedRenderOrder then
+		return cachedRenderOrder
+	end
+
+	local newRenderOrder = {}
+
+	for identifier, dataTable in pairs(elements) do
+		if dataTable.type == "Model" then
+			table.insert(newRenderOrder, 1, identifier)
+		elseif dataTable.type == "Sprite" or dataTable.type == "Quad" then
+			table.insert(newRenderOrder, identifier)
+		end
+	end
+
+	return newRenderOrder
+end
+
+---
+-- Gets the orientation of a given bone.
```

Comment:
Describe that it ignores the bone override etc. when a "rel" value is found.
Tbh. i still don't know how someone would use this or why